docs/designs/master_redesign.md

```diff
@@ -0,0 +1,136 @@
+# Master Design
+
+This document deseribes an ElasticDL master design.
+
+## Motivation
+
+We plan to provide elastic training feature to customized AllReduce training codes and other deep learning frameworks. Currently the master is designed for ElasticDL's own training framework. This requires us to re-design the master.
+After this re-design, the master consists of a few modular components and only launches those required components when providing elastic training feature to others.
+
+## Master Components
+The master has four modular components: elastic pod manager, task manager, rendezvous server and ElasticDL training/evaluation service.
+
+### Elastic Pod Manager
+The elastic pod manager is responsible for:
+
+1. Create PS/worker pods;
+2. Monitor pod events, keep track all pods status;
+3. Relaunch failed worker pods when needed.
+
+### Task Manager
+
+The task manager is responsible for creating tasks and dispatching tasks to workers. Workers will ask the task manager for new tasks and report task completion status to the task manager when needed. Optionally, the task mamager supports task fault-tolerance by recovering un-completed tasks from failed workers and re-assigning to other workers.
+
+For training jobs, the task manager supports dynamic sharding and partitions all training data into shards. Each task contains a data shard.
+
+### Rendezvous Server
+
+The rendezvous server supports for elastic AllReduce. After detecting worker pod number change, it creates a new worker host plan so that worker pods can re-initialize Horovod.
+
+### ElasticDL Training/Evaluation Service
+All other current master functions except the above three components. ElasticDL training/evaluation service supports for ElasticDL native training. It includes evaluation service, supporting for model saving and checkpoint, etc. 
+
+
+Thus, we can launch the master in seven component combinations:
```

Comment:
How do we specify which components we should launch? Use a command arguments?

Reply:
As mentioned in Implementation, we will add more arguments to specify which components are needed.